Bump LLVM to v19.1.7+1

Zentrik opened the pull request:
Including #55650 till that's merged.

Zentrik committed:
Bump LLVM to v19.1.1+0

Zentrik committed:
Update checksums

Zentrik commented:
The `/home/rag/Documents/Code/julia/usr/include/llvm/ADT/StringRef.h:871:20: warning: 'int __builtin_memcmp_eq(const void*, const void*, long unsigned int)' specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overread]` warning seems to be a gcc bug, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101361

giordano commented:
In that case you can guard the offending line with something like
```c
#pragma GCC diagnostic push
#if defined(_COMPILER_GCC_) && __GNUC__ >= 12 // if this is version-dependent
// Explain why this is being ignored...
#pragma GCC diagnostic ignored "-Wstringop-overflow"
#endif
...
#pragma GCC diagnostic pop
```

giordano commented:
Also, all tests are passing already on aarch64-darwin 🥳

Zentrik commented:
Looks like the only two issues are a incorrect warning during the build (mose's suggestion doesn't seem to work though `#pragma GCC diagnostic ignored "-Wstringop-overflow"` by itself does) and `/cache/build/tester-demeter6-13/julialang/julia-master/tmp/test-asan/asan/usr/bin/julia: error while loading shared libraries: libclang_rt.asan-x86_64.so: cannot open shared object file: No such file or directory`

Zentrik commented:
Guessing the asan problem is `clang -print-runtime-dir` is returning `toolchain/usr/lib/clang/19/lib/x86_64-unknown-linux-gnu` instead of `toolchain/usr/lib/clang/19/lib/linux` so we don't find `libclang_rt.asan...` to copy to the build folder.

giordano commented:
> though `#pragma GCC diagnostic ignored "-Wstringop-overflow"` by itself does

Without the push-and-pop, that becomes closer to just adding `-Wstringop-overflow` to
https://github.com/JuliaLang/julia/blob/80e60c89bb2b4fb193000c78e7f056c303dc6859/src/Makefile#L23-L25 (please always add a comment to explain why the warning is being skipped) but it'd be nicer to keep the ignore as local as possible, if feasible.

Zentrik committed:
Hack around change to `clang -print-runtime-dir`

`clang -print-runtime-dir` reports a non-existent directory as we build with `LLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF`. See https://github.com/llvm/llvm-project/pull/102834. I suspect https://github.com/llvm/llvm-project/commit/b6a1473f97d38471b08331dee7ae7f6112c495c0 caused the change by chaning the code in Driver.cpp to not check whether the printed directory existed.

Zentrik committed:
Only throw error when installing sanitizers

Zentrik committed:
Disable incorrect gcc warning

Locally I get:
```
In file included from /home/rag/Documents/Code/julia/usr/include/llvm/Object/ObjectFile.h:18,
                 from /home/rag/Documents/Code/julia/usr/include/llvm/DebugInfo/DIContext.h:18,
                 from /home/rag/Documents/Code/julia/src/debuginfo.cpp:6:
In function 'bool llvm::operator==(llvm::StringRef, llvm::StringRef)',
    inlined from 'bool llvm::operator!=(llvm::StringRef, llvm::StringRef)' at /home/rag/Documents/Code/julia/usr/include/llvm/ADT/StringRef.h:874:71,
    inlined from 'objfileentry_t find_object_file(uint64_t, llvm::StringRef)' at /home/rag/Documents/Code/julia/src/debuginfo.cpp:948:43,
    inlined from 'bool jl_dylib_DI_for_fptr(size_t, llvm::object::SectionRef*, int64_t*, llvm::DIContext**, bool, bool*, uint64_t*, void**, char**, char**)' at /home/rag/Documents/Code/julia/src/debuginfo.cpp:1135:34:
/home/rag/Documents/Code/julia/usr/include/llvm/ADT/StringRef.h:871:20: warning: 'int __builtin_memcmp_eq(const void*, const void*, long unsigned int)' specified bound 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overread]
  871 |     return ::memcmp(LHS.data(), RHS.data(), LHS.size()) == 0;
      |            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/rag/Documents/Code/julia/src/debuginfo.cpp: In function 'bool jl_dylib_DI_for_fptr(size_t, llvm::object::SectionRef*, int64_t*, llvm::DIContext**, bool, bool*, uint64_t*, void**, char**, char**)':
/home/rag/Documents/Code/julia/src/debuginfo.cpp:1133:11: note: source object allocated here
 1133 |     fname = dlinfo.dli_fname;
 ```

 On CI:
```
In file included from /cache/build/builder-amdci4-2/julialang/julia-master/usr/include/llvm/Object/ObjectFile.h:18,
                 from /cache/build/builder-amdci4-2/julialang/julia-master/usr/include/llvm/DebugInfo/DIContext.h:18,
                 from /cache/build/builder-amdci4-2/julialang/julia-master/src/debuginfo.cpp:6:
In function 'bool llvm::operator==(llvm::StringRef, llvm::StringRef)',
    inlined from 'bool llvm::operator!=(llvm::StringRef, llvm::StringRef)' at /cache/build/builder-amdci4-2/julialang/julia-master/usr/include/llvm/ADT/StringRef.h:874:71,
    inlined from 'objfileentry_t find_object_file(uint64_t, llvm::StringRef)' at /cache/build/builder-amdci4-2/julialang/julia-master/src/debuginfo.cpp:943:43,
    inlined from 'bool jl_dylib_DI_for_fptr(size_t, llvm::object::SectionRef*, int64_t*, llvm::DIContext**, bool, bool*, uint64_t*, void**, char**, char**)' at /cache/build/builder-amdci4-2/julialang/julia-master/src/debuginfo.cpp:1126:47:
/cache/build/builder-amdci4-2/julialang/julia-master/usr/include/llvm/ADT/StringRef.h:871:20: error: 'int __builtin_memcmp_eq(const void*, const void*, long unsigned int)' specified size 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
  871 |     return ::memcmp(LHS.data(), RHS.data(), LHS.size()) == 0;
      |            ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Zentrik committed:
Fix previous commit on gcc 9 and 10

Zentrik commented:
Looks like remaining issue is some failures in `analyzegc` that I guess weren't caught by previous versions of clang.

Zentrik commented:
First analyzegc failure looks incorrect, it seems to incorrectly think `uv_dup` could return 0 when `dupfd` is set to `-1` (minimal reproducer https://godbolt.org/z/4Wo99v1nv, https://github.com/llvm/llvm-project/issues/43015 looks to be the same issue).

Zentrik commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

giordano commented:
Need to rebase on master now that #56133 has been merged.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2409013158) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/fec1e8e_vs_3db1c62/report.md).

vtjnash commented:
> First analyzegc failure looks incorrect, it seems to incorrectly think `uv_dup` could return 0 when `dupfd` is set to `-1` (minimal reproducer https://godbolt.org/z/4Wo99v1nv, [llvm/llvm-project#43015](https://github.com/llvm/llvm-project/issues/43015) looks to be the same issue).

FWIW, the standard way to deal with this is to add an assert and a comment, as that will make both reviewers and bots happy

Zentrik commented:
The `union.array.(perf_sum, Float32, 0)` regression is due to SIMD not occurring. This is due to a change in the result from the instcombine pass. See https://godbolt.org/z/qqdzM7oxh,
The change from 
```
%12 = select i1 %exactly_isa.not, float -0.000000e+00, float %immutable_union.sroa.0.0.copyload, !dbg !77
%value_phi6 = fadd float %value_phi, %12, !dbg !77
```
to 
```
%12 = fadd float %value_phi, %immutable_union.sroa.0.0.copyload, !dbg !77
%value_phi6 = select i1 %exactly_isa.not, float %value_phi, float %12, !dbg !77
```
prevents the loopvectorize pass from SIMDing the code.

I suspect the other regression for the union.array benchmarks are similar.

Zentrik commented:
I'll try rebasing on top of https://github.com/JuliaLang/julia/pull/52850 and see if that fixes the regressions.

Zentrik commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2418652738) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e12026f_vs_35bf824/report.html) is available.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2422589980) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5cb1209_vs_bafd135/report.md).

oscardssmith commented:
looks like there are a few 8x (vectorization probably) failures here.

Zentrik commented:
PkgEval failure for McCormick is likely https://github.com/llvm/llvm-project/issues/113810

Zentrik committed:
Get `analyzegc` pass passing

The change to `GCChecker.cpp` is so the static analyzer doesn't see e.g. a call to a function pointer in llvm and then complain that it might be a safepoint.

Zentrik committed:
Make comment clearer

Zentrik committed:
Bump version bound

e7698a13e319a9919af04d3d693a6f6ea7168a44 isn't in llvm 19

giordano commented:
Analyzegc [is still failing](https://buildkite.com/julialang/julia-master/builds/41886#0192f1d7-2245-433e-9f2b-caf9636c7659/920-1719):
```
/cache/build/builder-demeter6-5/julialang/julia-master/usr/include/llvm/ADT/APInt.h:172:7: error: Attempt to free released memory [cplusplus.NewDelete]
  172 |       delete[] U.pVal;
      |       ^
```

Zentrik committed:
Bump to v19.1.1+1

Zentrik committed:
Bump LLVM tag to v19.1.1+1

Zentrik commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2496036348) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/128354f_vs_0bedaae/report.html) is available.

giordano commented:
On this PR `code_native(...; debuginfo=:none)` doesn't hide some debug info:
```julia-repl
julia> code_llvm(+, NTuple{2,Float64}; debuginfo=:none)
```
```llvm
; Function Signature: +(Float64, Float64)
define double @"julia_+_1011"(double %"x::Float64", double %"y::Float64") #0 {
top:
    #dbg_value(double %"x::Float64", !3, !DIExpression(), !16)
    #dbg_value(double %"y::Float64", !15, !DIExpression(), !16)
  %0 = fadd double %"x::Float64", %"y::Float64"
  ret double %0
}
```
Looks like the syntax of the debug values was changed.

giordano commented:
Another issue is that SVE vectorisation on aarch64 seems to be broken (in the sense that it isn't used).  Consider for example the function
```julia
function axpy!(a, x, y)
    @simd for idx in eachindex(x, y)
        @inbounds y[idx] = muladd(a, x[idx], y[idx])
    end
end
```
With `code_llvm(axpy!, (Float64, Vector{Float64}, Vector{Float64}); debuginfo=:none)` or `code_native(axpy!, (Float64, Vector{Float64}, Vector{Float64}); debuginfo=:none)` one case see that SVE registries aren't used anymore, but only NEON ones.  Performance isn't necessarily worse for same register width, but SVE has (potentially) larger width than NEON and is in general more capable (it supports more datatypes).

vtjnash commented:
IIUC, someone deleted support for printing comments on our end:
```
            std::unique_ptr<MCStreamer> S(TM->getTarget().createAsmStreamer(
#if JL_LLVM_VERSION >= 190000
                *Context, std::move(FOut), InstPrinter, std::move(MCE), std::move(MAB)
#else
                *Context, std::move(FOut), true, true, InstPrinter, std::move(MCE),
                std::move(MAB), false
#endif
                    ));
```
This is now supposed to be set here rather than deleted:
https://llvm.org/doxygen/classllvm_1_1MCTargetOptions.html#a192e9c2a63352ff1000ebe757e5b1f12

vtjnash commented:
MWE for the apparent clang-analyzer mistake:
```
130 vtjnash@deepsea4:~/julia$ clang++ -fPIE -std=c++20 dtor-test.cpp  && ./a.out                                                                                                                                       
C Ctor                                                                                                     
C Dtor                                                                                                     
vtjnash@deepsea4:~/julia$ g++ -fPIE -std=c++20 dtor-test.cpp  && ./a.out                                                                                                                                               
C Ctor                                                                                                                                                                                                                 
C Dtor                                                                                                     
$ cat dtor-test.cpp 
#include <iostream>
#include <optional>

class C {
  int *once;
public:
  C(int) {
    std::cout << "C Ctor" << std::endl;
    once = new int[3];
  }
  C(C&&) = delete;
  ~C() {
    std::cout << "C Dtor" << std::endl;
    delete[] once;
  }
};

int main() {
  std::optional<C> S{1};
  return 0;
}

$ ./usr/tools/clang++ --analyze -analyzer-output=text -std=c++20 dtor-test.cpp 
dtor-test.cpp:14:5: warning: Attempt to free released memory [cplusplus.NewDelete]                                                                                                                                     
   14 |     delete[] once;                                                                                 
      |     ^                                                                                                                                                                                                          
```

Zentrik commented:
Thanks for the patch but for reasons I don't understand yet, the following doesn't seem to work as debug info is not printed
```cpp
@@ -1274,8 +1276,13 @@ jl_value_t *jl_dump_function_asm_impl(jl_llvmf_dump_t* dump, char emit_mc, const
                 OutputAsmDialect = 1;
             MCInstPrinter *InstPrinter = TM->getTarget().createMCInstPrinter(
                 jl_ExecutionEngine->getTargetTriple(), OutputAsmDialect, MAI, MII, MRI);
-             std::unique_ptr<MCAsmBackend> MAB(TM->getTarget().createMCAsmBackend(
-                STI, MRI, TM->Options.MCOptions));
+            MCTargetOptions Options = TM->Options.MCOptions;
+            Options.AsmVerbose = true;
+            Options.MCUseDwarfDirectory = MCTargetOptions::EnableDwarfDirectory;
+            if (binary)
+                Options.ShowMCEncoding = true;
+            std::unique_ptr<MCAsmBackend> MAB(TM->getTarget().createMCAsmBackend(
+                STI, MRI, Options));
             std::unique_ptr<MCCodeEmitter> MCE;
             if (binary) { // enable MCAsmStreamer::AddEncodingComment printing
                 MCE.reset(TM->getTarget().createMCCodeEmitter(MII, *Context));
```
Instead if I do the following, the debug info is printed correctly
```cpp
diff --git a/src/jitlayers.cpp b/src/jitlayers.cpp
index e5a96cdf9d..a1fea54a16 100644
--- a/src/jitlayers.cpp
+++ b/src/jitlayers.cpp
@@ -1381,6 +1381,8 @@ namespace {
         options.MCOptions.ABIName = "lp64";
 #endif
 #endif
+        options.MCOptions.AsmVerbose = true;
+        options.MCOptions.MCUseDwarfDirectory = MCTargetOptions::EnableDwarfDirectory;
         uint32_t target_flags = 0;
         auto target = jl_get_llvm_target(imaging_default(), target_flags);
         auto &TheCPU = target.first;
```

Also, the `; Function Signature: +(Float64, Float64)` in missing in the output of `code_native` but not `code_llvm` for some other reason I haven't quite figured out.

giordano commented:
If that helps, the signature was added with  #50630, to give an idea of where to look at in the code.

Zentrik commented:
Thanks. On the function signature not printing it looks like we null out the printer here which prevents the printing
https://github.com/llvm/llvm-project/blob/9cdb7d2b6c333874ec969ef6ac64e0354bb3aa91/llvm/lib/CodeGen/AsmPrinter/DebugHandlerBase.cpp#L105-L108

giordano commented:
@Zentrik what's needed to move this forward?  It looks like you at least sorted the [debug printing](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2535302645) and had a decent idea of where the function signature was getting lost.

I'm concerned about [SVE not working](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2512322366) on aarch64 (interestingly enough, `vscale` vectorisation _does_ work on riscv64 with vector 1.0 extension), but don't know how to debug that, if anyone has hints please speak up!

Zentrik commented:
I'm largely stuck on getting the function signature printing working. I'm sure what can be done here, it looks like the rename of `AsmPrinterHandler` to `DebugHandlerBase` in https://github.com/llvm/llvm-project/commit/117b53ae38428ca66eaa886fb432e6f09db88fe4 actually changed the behaviour a bit. If someone more knowledgeable than me at LLVM once to take a look that would be great.

Is there any particular rush getting this in? I was assuming this would go in at the earliest after 1.12 is branched off.

giordano commented:
> Is there any particular rush getting this in? I was assuming this would go in at the earliest after 1.12 is branched off.

No particular rush related to v1.12 (it's not going to make it at this point), but I wanted to get a sense of where we're stuck and see what people can help with.

vtjnash committed:
some code_native fixes

vtjnash commented:
Rebased and added your fixes with mine. The "Function Signature" text is still missing which I couldn't investigate today

giordano commented:
Regarding the SVE issue, it may be due to LLVM optimization passes preferring NEON over SVE for some reason when targeting Neoverse V2 (which has 128-bit SVE registers, same width as NEON):

* [(unoptimised) LLVM IR on master, targeting `neoverse-v2`](https://godbolt.org/z/fdqEe66sh) (commit 8bf280277db)
* [(unoptimised) LLVM IR with this PR, targeting `neoverse-v2`](https://godbolt.org/z/hes18qjcz) (commit a90df10afca)

With both IRs, `opt` v18.1.0 always uses `vscale`, `opt` v19.1.0 always uses fixed-width `<2 x double>`.

However, for the same input, when targeting A64FX (which has 512-bit SVE registers) instead of Neoverse V2, `vscale` is always used:

* [(unoptimised) LLVM IR on master, targeting `a64fx`](https://godbolt.org/z/6vMree161) (commit 8bf280277db)
* [(unoptimised) LLVM IR with this PR, targeting `a64fx`](https://godbolt.org/z/a3aea5fbM) (commit a90df10afca)

Thanks @gbaraldi for suggesting doing this comparison between targets with different SVE register widths.

giordano commented:
Finally, to reassure myself that things still work, I tried a complex dot product on half precision, for which LLVM prefers SVE over NEON also on neoverse-v2:
```julia-repl
julia> function complex_dot(a::Vector{T}, b::Vector{T}) where {T}
           out = zero(eltype(a))
           for idx in eachindex(a, b)
               @fastmath out += conj(a[idx]) * b[idx]
           end
           out
       end
complex_dot (generic function with 1 method)

julia> code_llvm(complex_dot, NTuple{2,Vector{ComplexF16}}; debuginfo=:none)
```
```llvm
[...]
vector.body:                                      ; preds = %vector.body, %vector.ph
  %index = phi i64 [ 0, %vector.ph ], [ %index.next, %vector.body ]
  %vec.phi = phi half [ 0xH0000, %vector.ph ], [ %38, %vector.body ]
  %vec.phi92 = phi half [ 0xH0000, %vector.ph ], [ %40, %vector.body ]
  %offset.idx = shl i64 %index, 2
  %6 = or disjoint i64 %offset.idx, 4
  %7 = shl i64 %4, 5
  %8 = add i64 %7, %6
  %9 = getelementptr i8, ptr %memoryref_data, i64 %6
  %10 = getelementptr i8, ptr %memoryref_data, i64 %8
  %11 = getelementptr i8, ptr %9, i64 -4
  %12 = getelementptr i8, ptr %10, i64 -4
  %wide.vec = load <vscale x 16 x half>, ptr %11, align 2
  %wide.vec93 = load <vscale x 16 x half>, ptr %12, align 2
  %strided.vec = call { <vscale x 8 x half>, <vscale x 8 x half> } @llvm.vector.deinterleave2.nxv16f16(<vscale x 16 x half> %wide.vec)
  %13 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec, 0
  %14 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec, 1
  %strided.vec94 = call { <vscale x 8 x half>, <vscale x 8 x half> } @llvm.vector.deinterleave2.nxv16f16(<vscale x 16 x half> %wide.vec93)
  %15 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec94, 0
  %16 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec94, 1
  %17 = getelementptr i8, ptr %memoryref_data14, i64 %6
  %18 = getelementptr i8, ptr %memoryref_data14, i64 %8
  %19 = getelementptr i8, ptr %17, i64 -4
  %20 = getelementptr i8, ptr %18, i64 -4
  %wide.vec95 = load <vscale x 16 x half>, ptr %19, align 2
  %wide.vec96 = load <vscale x 16 x half>, ptr %20, align 2
  %strided.vec97 = call { <vscale x 8 x half>, <vscale x 8 x half> } @llvm.vector.deinterleave2.nxv16f16(<vscale x 16 x half> %wide.vec95)
  %21 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec97, 0
  %22 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec97, 1
  %strided.vec98 = call { <vscale x 8 x half>, <vscale x 8 x half> } @llvm.vector.deinterleave2.nxv16f16(<vscale x 16 x half> %wide.vec96)
  %23 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec98, 0
  %24 = extractvalue { <vscale x 8 x half>, <vscale x 8 x half> } %strided.vec98, 1
  %25 = fmul <vscale x 8 x half> %13, %21
  %26 = fmul <vscale x 8 x half> %15, %23
  %27 = fmul <vscale x 8 x half> %14, %22
  %28 = fmul <vscale x 8 x half> %16, %24
  %29 = fadd <vscale x 8 x half> %25, %27
  %30 = fadd <vscale x 8 x half> %26, %28
  %31 = fmul <vscale x 8 x half> %13, %22
  %32 = fmul <vscale x 8 x half> %15, %24
  %33 = fmul <vscale x 8 x half> %14, %21
  %34 = fmul <vscale x 8 x half> %16, %23
  %35 = fsub <vscale x 8 x half> %31, %33
  %36 = fsub <vscale x 8 x half> %32, %34
  %37 = call half @llvm.vector.reduce.fadd.nxv8f16(half %vec.phi, <vscale x 8 x half> %29)
  %38 = call half @llvm.vector.reduce.fadd.nxv8f16(half %37, <vscale x 8 x half> %30)
  %39 = call half @llvm.vector.reduce.fadd.nxv8f16(half %vec.phi92, <vscale x 8 x half> %35)
  %40 = call half @llvm.vector.reduce.fadd.nxv8f16(half %39, <vscale x 8 x half> %36)
  %index.next = add nuw i64 %index, %5
  %41 = icmp eq i64 %index.next, %n.vec
  br i1 %41, label %middle.block, label %vector.body
[...]
```
```julia-repl
julia> code_native(complex_dot, NTuple{2,Vector{ComplexF16}}; debuginfo=:none)
```
```as
.LBB0_6:                                // %vector.body
                                        // =>This Inner Loop Header: Depth=1
        ld2h    { z2.h, z3.h }, p0/z, [x15]
        ld2h    { z4.h, z5.h }, p0/z, [x15, #2, mul vl]
        addvl   x15, x15, #4
        addvl   x14, x14, #-1
        ld2h    { z6.h, z7.h }, p0/z, [x16]
        fmul    z18.h, z2.h, z6.h
        fmul    z20.h, z3.h, z7.h
        ld2h    { z16.h, z17.h }, p0/z, [x16, #2, mul vl]
        fmul    z19.h, z4.h, z16.h
        fmul    z21.h, z5.h, z17.h
        addvl   x16, x16, #4
        fadd    z18.h, z18.h, z20.h
        fmul    z20.h, z2.h, z7.h
        fmul    z2.h, z3.h, z6.h
        fadd    z19.h, z19.h, z21.h
        fmul    z21.h, z4.h, z17.h
        fmul    z3.h, z5.h, z16.h
        fadda   h0, p0, h0, z18.h
        fsub    z2.h, z20.h, z2.h
        fsub    z3.h, z21.h, z3.h
        fadda   h1, p0, h1, z2.h
        fadda   h0, p0, h0, z19.h
        fadda   h1, p0, h1, z3.h
        cbnz    x14, .LBB0_6
// %bb.7:                               // %middle.block
        cbz     x13, .LBB0_10
```
Note the operations on ` <vscale x 8 x half>` in LLVM IR, and on `z` registers in native code.

I guess the takeaway is that LLVM changed some heuristics to prefer SVE vs NEON when targeting CPUs with 128-bit SVE registers (same as NEON).

giordano committed:
Fix llvmunwind checksums

vtjnash commented:
> I'm sure what can be done here, it looks like the rename of AsmPrinterHandler to DebugHandlerBase in https://github.com/llvm/llvm-project/commit/117b53ae38428ca66eaa886fb432e6f09db88fe4 actually changed the behaviour a bit

Yep, looks like this is an implementation issue on llvm's side introduced by that commit. I've made a PR request to undo the breaking parts of that change and restore the functionality and unnecessary API breakage there: https://github.com/llvm/llvm-project/pull/122297

We can backport that to our llvm 19 branch (with some minor conflicts), or simply have that be broken until llvm 20 is released.

giordano commented:
To summarise:

* code native/llvm printing is mostly fixed
* only remaining issue is printing of the "Function Signature" line with possible solutions outlined at https://github.com/JuliaLang/julia/pull/56130#issuecomment-2580552182
* SVE on neoverse-v2 turned out to be a [non-issue](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2579817064) (apparently it's a change of behaviour in LLVM, and that's consistent with [Clang which uses NEON already in previous versions](https://godbolt.org/z/cK8srdzs4) for the same target
* all Julia tests are passing on CI here
* most of the checks are passing, including llvmpasses
* only failure is analyzegc, for which Jameson opened an issue upstream: https://github.com/llvm/llvm-project/issues/119415 (but we still need to resolve it here)

Should we run PkgEval tests again?  The previous logs seem to be gone now.

Zentrik commented:
> Should we run PkgEval tests again? The previous logs seem to be gone now.

I think it would be best to wait till this PR has the rest of the fixes in, but no harm in doing so.

oscardssmith commented:
presumably the fix for the analyzeGC issue is to backport https://github.com/llvm/llvm-project/pull/122330

Zentrik committed:
Try workarounding clang static analyzer bug

Zentrik committed:
Revert "Try workarounding clang static analyzer bug"

This reverts commit 2252e894370ff5bad11613771657cd244431bd17.

gbaraldi committed:
Address clangsa concerns on macos + add workaround for static analyzer bug

Zentrik committed:
Remove ineffective workaround for clangsa bug

Zentrik committed:
Adapt to https://github.com/llvm/llvm-project/commit/f6b0555a433cea1d32a6904c120516cd94b8f3db

Zentrik committed:
Bump LLVM to v19.1.7+0

Zentrik committed:
Refresh checksums

Zentrik commented on src/jitlayers.cpp:
```diff
@@ -1258,7 +1258,7 @@ class JLMemoryUsagePlugin : public ObjectLinkingLayer::Plugin {
             (void) code_size;
             (void) data_size;
-            this->jit_bytes_size.fetch_add(graph_size, std::memory_order_relaxed);
+            jl_atomic_fetch_add_relaxed(this->jit_bytes_size, graph_size)
```
```suggestion
            jl_atomic_fetch_add_relaxed(this->jit_bytes_size, graph_size);
```

Zentrik committed:
Update src/jitlayers.cpp

Zentrik commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Zentrik commented:
@nanosoldier `runtests()`

Zentrik commented:
```
Error in testset cmdlineargs:
Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-1/julialang/julia-master/julia-b71c902cd0/share/julia/test/cmdlineargs.jl:1092
  Expression: occursin(r"\.jl:(\d+)", bt)
   Evaluated: occursin(r"\.jl:(\d+)", "\n[3654] signal 10 (3): Bus error\nin expression starting at none:0\njl_valid_type_param at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-3/julialang/julia-master/src/builtins.c:1540\nunknown function (ip: 0x85426410f) at (unknown file)\nAllocations: 1 (Pool: 1; Big: 0); GC: 0")
```

```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-d5c0a02be5/share/julia/test/cmdlineargs.jl:1092
  Expression: occursin(r"\.jl:(\d+)", bt)
   Evaluated: occursin(r"\.jl:(\d+)", "\n[180] signal 11 (2): Segmentation fault: 11\nin expression starting at none:0\njl_valid_type_param at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1540\njl_f_apply_type at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1587\nAllocations: 1 (Pool: 1; Big: 0); GC: 0")
```
Seem concerning

Zentrik commented:
```
Error in testset cmdlineargs:
Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-1/julialang/julia-master/julia-b71c902cd0/share/julia/test/cmdlineargs.jl:1092
  Expression: occursin(r"\.jl:(\d+)", bt)
   Evaluated: occursin(r"\.jl:(\d+)", "\n[3654] signal 10 (3): Bus error\nin expression starting at none:0\njl_valid_type_param at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-3/julialang/julia-master/src/builtins.c:1540\nunknown function (ip: 0x85426410f) at (unknown file)\nAllocations: 1 (Pool: 1; Big: 0); GC: 0")
```
```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-d5c0a02be5/share/julia/test/cmdlineargs.jl:1092
  Expression: occursin(r"\.jl:(\d+)", bt)
   Evaluated: occursin(r"\.jl:(\d+)", "\n[180] signal 11 (2): Segmentation fault: 11\nin expression starting at none:0\njl_valid_type_param at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1540\njl_f_apply_type at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1587\nAllocations: 1 (Pool: 1; Big: 0); GC: 0")
```

Seems concerning

giordano commented:
Uhm, tests were passing in https://buildkite.com/julialang/julia-master/builds/43624 (https://github.com/JuliaLang/julia/commit/c5266cb3208056cfb2ff8820b2c0b358eaf607ea), apart from the analyzegc check. At least that one is working now 😅

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2652137791) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/2a4bc6d_vs_1de5880/report.md).

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2652138098) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/b71c902_vs_1de5880/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**7 packages crashed only on the current version.**

 - Invalid LLVM IR was generated: 3 packages
 - An internal error was encountered: 2 packages
 - A segmentation fault happened: 2 packages

110 packages crashed on the previous version too.

## ✖ Packages that failed

**12 packages failed only on the current version.**

 - Package fails to precompile: 1 packages
 - Illegal method overwrites during precompilation: 1 packages
 - Package has test failures: 2 packages
 - Package tests unexpectedly errored: 1 packages
 - Tests became inactive: 2 packages
 - Test duration exceeded the time limit: 5 packages

1941 packages failed on the previous version too.

## ✔ Packages that passed tests

**13 packages passed tests only on the current version.**

 - Other: 13 packages

4732 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**2 packages successfully loaded only on the current version.**

 - Other: 2 packages

2559 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

**2 packages were skipped only on the current version.**

 - Package could not be installed: 2 packages

918 packages were skipped on the previous version too.



</details>

Zentrik commented:
NIfTI failure is a llvm bug, https://github.com/llvm/llvm-project/issues/127128.

Zentrik commented:
> https://buildkite.com/julialang/julia-master/builds/44665#0194f6b4-fff4-4106-a917-6a8526bf6cf8/764-1551
> 
> ```
> The global RNG seed was 0x76717190b3fd71699453c7b6416351f.
> Error in testset cmdlineargs:
> Test Failed at /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-1/julialang/julia-master/julia-b71c902cd0/share/julia/test/cmdlineargs.jl:1092
>   Expression: occursin(r"\.jl:(\d+)", bt)
>    Evaluated: occursin(r"\.jl:(\d+)", "\n[3654] signal 10 (3): Bus error\nin expression starting at none:0\njl_valid_type_param at /usr/home/julia/.buildkite-ag
> ent/builds/freebsd13-amdci6-3/julialang/julia-master/src/builtins.c:1540\nunknown function (ip: 0x85426410f) at (unknown file)\nAllocations: 1 (Pool: 1; Big: 0)
> ; GC: 0")
> ```
> 
> https://buildkite.com/julialang/julia-master/builds/44650#0194f563-4418-4052-9f81-ba96a20b77b6/1195-2143
> 
> ```
> Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-HL2F7YQ3XH.0/build/default-honeycrisp-HL2F7YQ3XH-0/julialang/julia-master/julia-d5c0a02be5/share/julia/test/cmdlineargs.jl:1092
>   Expression: occursin(r"\.jl:(\d+)", bt)
>    Evaluated: occursin(r"\.jl:(\d+)", "\n[180] signal 11 (2): Segmentation fault: 11\nin expression starting at none:0\njl_valid_type_param at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1540\njl_f_apply_type at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-XC9YQX9HH2.0/build/default-honeycrisp-XC9YQX9HH2-0/julialang/julia-master/src/builtins.c:1587\nAllocations: 1 (Pool: 1; Big: 0); GC: 0")
> ```
> 
> Seems concerning

Ok these don't seem to be llvm 19 issues, similar problem here https://buildkite.com/julialang/julia-master/builds/44889#01951472-9479-4ffc-bbdc-470530837d24/1091-1402. On slack, Jameson says this is possibly due to CI running with multiple threads.

Zentrik committed:
Bump to 19.1.7+1

Zentrik commented on stdlib/libLLVM_jll/Project.toml:
```diff
@@ -1,13 +1,13 @@
 
 [compat]
-julia = "1.11"
+julia = "1.12"
```
Do we need to bump this to 1.13? Same q for lld_jll

giordano replied:
Maybe yes for consistency.

Zentrik replied:
```suggestion
julia = "1.13"
```

Zentrik committed:
Refresh checksums

I think they got screwed up because I ran them all once with many jobs or because I control Ced it a couple times, `make -f contrib/refresh_checksums.mk llvm clang lld -j256`.

Zentrik commented:
Given the recent compiler changes, https://github.com/JuliaLang/julia/pull/57389, https://github.com/JuliaLang/julia/pull/52850, curious to see what the benchmarks look like now

@nanosoldier `runbenchmarks(ALL, vs=":master")`

Zentrik committed:
Test reverting "Change memory indexing to use the type as index instead of i8 (#57389)"

This reverts commit b9a8d46a234fa0fe625371e107f3ab441a867182.

Zentrik committed:
Revert "Test reverting "Change memory indexing to use the type as index instead of i8 (#57389)""

This reverts commit 8e4e86da8d54d3c91ce20b739a5cc3b0977bd23a.

Zentrik committed:
Revert "Refresh checksums"

This reverts commit 0430aa776644d65feea69afe90875a5d503e76d0.

Zentrik committed:
Revert "Bump to 19.1.7+1"

This reverts commit 544c0c8f0a511aa69fc45674e38e18af0fc97775.

Zentrik commented:
Ahh I should have looked at the assert build, 
```
      From worker 17:	  call swiftcc void @julia_SubString_43862(ptr noalias nocapture noundef sret({ ptr addrspace(10), i64, i64 }) %10, ptr noalias nocapture noundef %262, ptr nonnull swiftself %2, ptr addrspace(10) %14, i64 signext %265, i64 signext %266), !dbg !68580
      From worker 17:	  %12 = alloca ptr addrspace(10), align 8
      From worker 17:	  %9 = alloca ptr addrspace(10), align 8
      From worker 17:	julia: /cache/build/tester-amdci5-13/julialang/julia-master/src/llvm-late-gc-lowering.cpp:1231: State LateLowerGCFrame::LocalScan(llvm::Function&): Assertion `false && "Expected single alloca"' failed.
      From worker 17:
      From worker 17:	[4232] signal 6 (-6): Aborted
      From worker 17:	in expression starting at none:1
      From worker 17:	Allocations: 632820359 (Pool: 632817861; Big: 2498); GC: 64
compileall                                       (17) |         failed at 2025-02-24T01:51:04.982
Test Failed at /cache/build/tester-amdci4-9/julialang/julia-master/julia-0430aa7766/share/julia/test/compileall.jl:6
  Expression: success(pipeline(`$(Base.julia_cmd()) --compile=all --strip-ir --output-o $(dir)/sys.o.a -e 'exit()'`, stderr = stderr))
```

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/56130#issuecomment-2677181050) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b00614a_vs_9294132/report.md).

gbaraldi commented:
I guess it deserves a look, but those benchmarks have been very brittle to me in the past. They require LLVM to move a lot of code around to vectorize

gbaraldi committed:
Fix late gc lowering issue

gbaraldi commented:
It's a straight up regression in LLVM https://godbolt.org/z/ohdYjz8xP

Zentrik commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -1219,22 +1219,18 @@ State LateLowerGCFrame::LocalScan(Function &F) {
-                                        SRet_gc = gc_allocas.pop_back_val();
+                                    if (gc_allocas.size() == 0) {
+                                        llvm_dump(CI);
+                                        assert(false && "Expected at least one alloca");
```
Is there a reason this is an assert instead of just printing to stderr, would be nice for the message to show up on non assert builds.

gbaraldi replied:
I guess this can be an abort

giordano commented:
@gbaraldi  https://buildkite.com/julialang/julia-master/builds/45109#01953924-b6fd-458a-8d78-259918cbf674/1250-1272
```
******************** TEST 'Julia :: late-lower-gc-sret.ll' FAILED ********************
Exit Code: 2
Command Output (stderr):
--
RUN: at line 3: opt --load-pass-plugin=libjulia-codegen.so -passes='function(LateLowerGCFrame)' -S /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll | FileCheck /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll
+ FileCheck /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll
+ opt --load-pass-plugin=libjulia-codegen.so '-passes=function(LateLowerGCFrame)' -S /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll
opt: /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll:53:13: error: invalid redefinition of function 'julia.get_pgcstack'
declare ptr @julia.get_pgcstack()
            ^
FileCheck error: '<stdin>' is empty.
FileCheck command line:  FileCheck /cache/build/builder-amdci4-5/julialang/julia-master/test/llvmpasses/late-lower-gc-sret.ll
--
********************
```

gbaraldi committed:
Remove double definition

gbaraldi commented:
Opened https://github.com/llvm/llvm-project/issues/128557. It looks like it's the result of https://github.com/llvm/llvm-project/pull/113686 which after some fixes made it so that this is a legal optimization under fastmath rules but not legal under normal FP

Zentrik committed:
Revert "Revert "Bump to 19.1.7+1""

This reverts commit b08d8419277d196a0dad8d534c41236c19b94063.

Zentrik committed:
Revert "Revert "Refresh checksums""

This reverts commit d428d02b4fc11a17a2d03f7a0b3553e4925815dc.

gbaraldi committed:
Change asserts to aborts

giordano committed:
Update src/llvm-late-gc-lowering.cpp

giordano commented on src/llvm-late-gc-lowering.cpp:
```diff
@@ -1138,11 +1139,13 @@ static SmallSetVector<AllocaInst *, 8> FindSretAllocas(Value* SRetArg) {
                     llvm_dump(SI);
-                    assert(false && "Malformed Select");
+                    dbgs() << "Malformed Select\n";
+                    abort;
```
```suggestion
                    abort();
```

giordano committed:
Update src/llvm-late-gc-lowering.cpp

Zentrik commented on stdlib/LLD_jll/Project.toml:
```diff
@@ -9,8 +9,8 @@ libLLVM_jll = "8f36deef-c2a5-5394-99ed-8e07531fb29a"
 [compat]
-julia = "1.11"
-libLLVM_jll = "18.1.7"
+julia = "1.12"
```
```suggestion
julia = "1.13"
```

Zentrik committed:
Bump compat to 1.13
